Linter CI part 1

yaswant opened the pull request:
# Description

## Summary

Prepare changes for #72 

## Changes

Added
- Update scripts to pass CI checks 

## Dependency

None.

## Impact

None.

## Issues addressed

Resolves

#51 

## Coordinated merge

#72


## Checklist

- [x] I have performed a self-review of my own changes

yaswant commented on .github/linters/.ruff.toml:
```diff
@@ -0,0 +1,21 @@
+cache-dir = "/tmp/.ruff_cache"
+line-length = 120
```
```suggestion
line-length = 80
```

yaswant commented on .github/linters/.ruff.toml:
```diff
@@ -0,0 +1,21 @@
+line-length = 120
+
+[lint]
+# C0114: Missing module docstring (missing-module-docstring)
```
```suggestion
```

yaswant commented on .github/linters/.ruff.toml:
```diff
@@ -0,0 +1,21 @@
+# C0209: Formatting a regular string which could be an f-string (consider-using-f-string)
+# W0622: Redefining built-in 'copyright' (redefined-builtin)
+# R0801: Similar lines in 2 files
+ignore = ["C0114","C0116","C0103","C0209","W0622","R0801",]
```
```suggestion
ignore = ["C0116","C0103","C0209","W0622","R0801",]
```

yaswant reviewed (commented):
some minor changes in ruff.toml

yaswant commented on .github/linters/.ruff.toml:
```diff
@@ -0,0 +1,20 @@
+cache-dir = "/tmp/.ruff_cache"
+line-length = 80
+
+[lint]
+# C0116: Missing function or method docstring (missing-function-docstring)
+# C0103: Constant name doesn't conform to UPPER_CASE naming style (invalid-name)
+# C0209: Formatting a regular string which could be an f-string (consider-using-f-string)
+# W0622: Redefining built-in 'copyright' (redefined-builtin)
+# R0801: Similar lines in 2 files
+ignore = ["C0116","C0103","C0209","W0622","R0801",]
```
```suggestion
```

yaswant commented:
- @yaswant to update shell scripts (Fixes #53), switch off ruff linting for now..
- @james-bruten-mo to blackify Python scripts

yaswant commented:
@james-bruten-mo Don't think there is anyway to override black defaults for SuperLinter.  I see now SciTools/iris switched from black to ruff.  So there must be a way AVD use the tool.  I take the message  that if you prefer black, then just don't try to override defaults via command-line.

Pierre-siddall commented on .github/linters/.ruff.toml:
```diff
@@ -0,0 +1,13 @@
+
+[format]
+# Like Black, indent with spaces, rather than tabs.
+indent-style = "space"
```
I really agree with the usage of spaces as a standard for indenting here as it will avoid the extra overhead involved in linting escape characters.

yaswant commented:
This is now ready for review and merge. @Pierre-siddall you can step away from the PR for now, but FYI we have disabled ruff in the workflow which will be revisited in later.

Requesting reviews from: 
- @james-bruten-mo to check implementation changes
- @r-sharp to check POSIX compliance in the script we discussed in teams chat
- @t00sa workflow implantation
Thanks

james-bruten-mo commented on kgo_updates/kgo_update/meto_run_kgo_script.sh:
```diff
@@ -20,145 +20,68 @@ GREEN='\033[0;32m'
 
-    # Run the Update Script
-    $kgo_command_spice
+    echo;echo "${GREEN}Running KGO Update Script on ${_platform}.${NC}"
```
`echo -ne`

yaswant replied:
Bulk of the change I made this morning to comply with POSIX! Unfortunately `-n` `-e` flags are not POSIX-compliant.

james-bruten-mo commented on kgo_updates/kgo_update/meto_run_kgo_script.sh:
```diff
@@ -20,145 +20,68 @@ GREEN='\033[0;32m'
-    scp -q kgo_update.py frum@$host_xc40:~
-
-    # Define the commands to run on xc40
-    command=". /etc/profile ; module load scitools ;
```
We need the `module load scitools` when on the xc40s to get python3. I guess there's no reason to not load scitools on all platforms, so probably easiest to just add this to the command on new line 61.

yaswant replied:
Fixed in https://github.com/MetOffice/SimSys_Scripts/pull/52/commits/3ddd0d1e620dece5d8130d308aa80ae3c1bcb5a4

james-bruten-mo commented on kgo_updates/kgo_update/meto_run_kgo_script.sh:
```diff
@@ -20,145 +20,68 @@ GREEN='\033[0;32m'
-        printf "${GREEN}Rsyncing the generated files into spice ${variables_dir}.\n${NC}"
-        if [ $new_release -ne 1 ]; then
+    if eval "$_command"; then
+        echo;echo "${GREEN}Successfully installed on ${_platform}.${NC}"
```
`echo -ne`

yaswant replied:
Not POSIX-compliant

james-bruten-mo commented on kgo_updates/kgo_update/meto_run_kgo_script.sh:
```diff
@@ -20,145 +20,68 @@ GREEN='\033[0;32m'
-        if [ $new_release -ne 1 ]; then
+    if eval "$_command"; then
+        echo;echo "${GREEN}Successfully installed on ${_platform}.${NC}"
+        echo "${GREEN}Moving the generated files into ${_platform} ${variables_dir}.${NC}"
```
`echo -ne`

yaswant replied:
same as above

james-bruten-mo commented on kgo_updates/kgo_update/meto_run_kgo_script.sh:
```diff
@@ -20,145 +20,68 @@ GREEN='\033[0;32m'
-        rsync --remove-source-files -avz \
-                umadmin@$host_ex:~/kgo_update_${new_kgo_dir}.sh \
-                ${variables_dir}/ex1a_update_script.sh
+        echo "${RED}The installation script has failed on ${_platform}.${NC}"
```
`echo  -ne`

yaswant replied:
You would think, its much better to write non-POSIX standard scripts these days.  gnu bash is present everywhere on the cloud these days.  I'll have a chat higher up.

james-bruten-mo replied:
Keeping the colours printing out as they currently do is definitely more important than keeping it posix compliant. I'd say we either go with `echo -ne` or we tell shell-check to ignore those errors and keep them as they were

yaswant replied:
Sure, I think I missed the trick there what I originally posted you on the chat. We can have both with:
```
RED=$(tput setaf 1)
GREEN=$(tput setaf 2)
NC=$(tput sgr0)

echo "${GREEN}I am green${NC} "
```

james-bruten-mo replied:
Yep, that looks like it works, cheers

james-bruten-mo reviewed (commented):
I've just used this branch for the kgo install I've been doing today. This highlighted a couple of issues:

* The xc40's require scitools in order to have python3 available - this causes f-string issues when running kgo_update.py - I think we can just add `module load scitools` to the command in the new function as there's no problem with having it loaded on other platforms.
* The new printf syntax doesn't colour the strings correctly - instead it just prints out the colour code string. I tested Sams fix from #53 which is to use `echo -ne` instead and that seems to work so suggest we go with that? I highlighted the locations in `meto_run_kgo_script.sh` but not in `meto_update_kgo.sh`. 

Cheers

james-bruten-mo commented on kgo_updates/kgo_update/meto_run_kgo_script.sh:
```diff
@@ -20,145 +20,68 @@ GREEN='\033[0;32m'
+}
+
+if [ "${platforms#*spice}" != "$platforms" ] && [ "${platforms#*azspice}" = "$platforms" ]; then
+    run_update_script "spice" "$kgo_command_spice" "frum"
```
Sorry, another point I've just noticed is what does `host` default above to for spice and azspice? 

And the user for azspice and ex1a should be `umadmin`

yaswant replied:
fixed in https://github.com/MetOffice/SimSys_Scripts/pull/52/commits/3ddd0d1e620dece5d8130d308aa80ae3c1bcb5a4

james-bruten-mo commented:
I've had another go at installing kgo with this branch and hit issues again.
Given those changes are becoming reasonably large, can I suggest that we split a refactor of the the kgo shell script off into a separate PR, and just enforce shellcheck compliance (probably with some exemptions) in this ticket? Good development practice would probably suggest this ticket should only be for achieving linter compliance anyway.
We can then test the install script changes more thoroughly in slower time.

yaswant commented:
> I've had another go at installing kgo with this branch and hit issues again. Given those changes are becoming reasonably large, can I suggest that we split a refactor of the the kgo shell script off into a separate PR, and just enforce shellcheck compliance (probably with some exemptions) in this ticket? Good development practice would probably suggest this ticket should only be for achieving linter compliance anyway. We can then test the install script changes more thoroughly in slower time.

I've reverted the script back to the way it was before with fixes against shellcheck. Please give it another go.

james-bruten-mo commented:
I've tested again and the printf statements with the colour codes at the end don't colour the line correctly. I've reverted these and added the relevant shellcheck ignores if that's an acceptable way to go?

yaswant commented:
@james-bruten-mo frustration won, eventually 😉

Pierre-siddall commented on styling_keywords.py:
```diff
@@ -9,222 +9,1426 @@
+    "then",
+    "this_image",
+    "threadprivate",
+    "time",
+    "time8",
+    "tiny",
+    "to",
+    "trailz",
+    "transfer",
+    "transpose",
+    "trim",
+    "true",
+    "ttynam",
+    "type",
+    "ubound",
+    "ucobound",
+    "umask",
+    "unformatted",
+    "unit",
+    "unlink",
+    "unlock",
+    "unpack",
+    "use",
+    "value",
+    "verif",
+    "verify",
+    "volatile",
+    "wait",
+    "where",
+    "while",
+    "workshare",
+    "write",
+    "xor",
+    "zabs",
+    "zcos",
+    "zexp",
+    "zlog",
+    "zsin",
+    "zsqrt",
+}
```
At some point when it's less high risk to do so it would be good to merge this list with the one above it to avoid any potential duplication.

Pierre-siddall commented on suite_report.py:
```diff
@@ -2061,8 +1976,7 @@ def print_report(self):
                 [
-                    "There has been an exception in "
-                    + "SuiteReport.print_report()",
+                    "There has been an exception in " + "SuiteReport.print_report()",
```
Unification on this string.

```suggestion
                    "There has been an exception in SuiteReport.print_report()",
```

Pierre-siddall reviewed (commented):
This is all good stuff and seems like it will run smoothly, I've added a few suggestions regarding reverts of black changes and a comment about the future unification of the keywords list which I believe we've already picked up on.

Pierre-siddall commented:
> @Pierre-siddall you can open a new issue/pr to address your suggestion. This PR has already got very crowded with multiple ideas which is a digression from the original purpose. Let's get the CI workflow in first.

@yaswant I'll probably add these suggestions into #70  as the unification of the kgo updates script became an also suggestion.

Pierre-siddall reviewed (approved):
The previously suggested changes are better suited to a new issue or PR. Therefore as the semantics of this code looks fine I'm happy to approve this PR 😄.

james-bruten-mo commented:
Part of the intention of black is that it forces you to do the styling a single way so everything is consistent. But the last string in the  suite_report suggestion is a good one to change as it won't combine strings that have already been split up so I've committed that. 
As Yash suggests, the keywords work should be it's own ticket

t00sa commented:
This PR is much to large and untidy and needs to be split up.  It should be fairly straight forward to separate the original workflow changes, and split off the various script changes into their own items.  This shouldn't greatly increase the workload if individual components have already been reviewed.

james-bruten-mo commented:
I've just deleted the .github directory so this is now only the styling changes. Will open a new PR for the CI in a sec...

yaswant commented:
@james-bruten-mo I was going to suggest other way around:, but since you have already done this, I suggest you bring back `README` changes to this branch

james-bruten-mo commented:
> @james-bruten-mo I was going to suggest other way around:, but since you have already done this, I suggest you bring back `README` changes to this branch

Yes, I was thinking that way, but the .github changes were a lot more self-contained so easier to move!
I'll bring back the readme though

t00sa reviewed (changes requested):
Almost all of this is fine.  A couple of minor fixes for places were black hasn't concatenated strings together properly.

t00sa reviewed (approved):
Thanks.  I'm happy to approve this
